spec/index.bs

```diff
@@ -6974,10 +7017,68 @@ interface mixin GPUProgrammablePassEncoder {
+
+    **Arguments:**
+    <dl>
+        : {{GPUProgrammablePassEncoder}} |encoder|
+        :: Encoder whose bind groups are being validated.
+        : {{GPUPipelineBase}} |pipeline|
+        :: Pipeline to validate |encoder|s bind groups are compatible with.
+    </dl>
+
+    1. For each |stage| in [{{GPUShaderStage/VERTEX}}, {{GPUShaderStage/FRAGMENT}}, {{GPUShaderStage/COMPUTE}}]:
+        1. Let |bufferBindings| be a [=list=] of ({{GPUBufferBinding}}, `boolean`) pairs,
+            where the latter indicates whether the resource was used as writable.
+        1. Let |textureViews| be a [=list=] of ({{GPUTextureView}}, `boolean`) pairs,
+            where the latter indicates whether the resource was used as writable.
+        1. For each pair of ({{GPUIndex32}} |index|, {{GPUBindGroupLayout}} |bindGroupLayout|) in
+            |pipeline|.{{GPUPipelineBase/[[layout]]}}.{{GPUPipelineLayout/[[bindGroupLayouts]]}}:
+            1. Let |bindGroupEntries| be
+                |encoder|.{{GPUProgrammablePassEncoder/[[bind_groups]]}}[|index|].{{GPUBindGroupDescriptor/entries}}.
+            1. Let |bindGroupLayoutEntries| be
+                |bindGroupLayout|.{{GPUBindGroupLayout/[[descriptor]]}}.{{GPUBindGroupLayoutDescriptor/entries}}.
+            1. For each {{GPUBindGroupEntry}} |bindGroupLayoutEntry| in |bindGroupLayoutEntries|
+                for which |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/visibility}} contains |stage|:
+                1. Let |bindGroupEntry| be the {{GPUBindGroupEntry}} in |bindGroupEntries| for which
+                    |bindGroupEntry|.{{GPUBindGroupEntry/binding}} is equal to
+                    |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/binding}}.
+                1. If |bindGroupEntry|.{{GPUBindGroupEntry/resource}} is a {{GPUBufferBinding}}:
+
+                    1. Let {{GPUBufferBinding}} |resource| be |bindGroupEntry|.{{GPUBindGroupEntry/resource}}.
+                    1. Let |resourceWritable| be (|bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/type}} == {{GPUBufferBindingType/"storage"}}).
+                    1. For each pair ({{GPUBufferBinding}} |pastResource|, `boolean` |pastResourceWritable|) in |bufferBindings|:
+                        1. If (|resourceWritable| or |pastResourceWritable|) is true, and
+                            |pastResource| and |resource| are [=buffer-binding-aliasing=], return `true`.
+                    1. [=map/set|Set=] |bufferBindings|[|resource|] to |resourceWritable|.
+
+                    Otherwise, if |bindGroupEntry|.{{GPUBindGroupEntry/resource}} is a {{GPUTextureView}}:
+
+                    1. Let {{GPUTextureView}} |resource| be |bindGroupEntry|.{{GPUBindGroupEntry/resource}}.
+                    1. Let |resourceWritable| be (|bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/storageTexture}}.{{GPUStorageTextureBindingLayout/access}} == {{GPUStorageTextureAccess/"write-only"}}).
```

Comment:
thank you for writing this in a generic way instead of relying on the fact we don't have read-only storage textures :)

Reply:
This is just a refactor from Myles's PR :) We're still going to have to edit it if we add read-write, unfortunately!